### PR TITLE
Coke oven and Blast furnace

### PR DESCRIPTION
**Pull request by LetterN**

### CODE REVIEW

File: src/main/java/mods/railcraft/world/item/crafting/RailcraftRecipeSerializers.java [hunk 1]
```diff
@@ -18,4 +18,7 @@
 
   public static final RegistryObject<IRecipeSerializer<?>> ROLLING =
       RECIPE_SERIALIZERS.register("rolling", RollingRecipe.RollingRecipeSerializer::new);
+
+  public static final RegistryObject<IRecipeSerializer<?>> COKEING =
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/item/crafting/RailcraftRecipeSerializers.java]: Call this coke_oven_cooking

File: src/main/java/mods/railcraft/world/level/block/CokeOvenBlock.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+import net.minecraft.block.Block;
+import net.minecraft.block.material.Material;
+import net.minecraftforge.common.ToolType;
+
+public class CokeOvenBlock extends Block {
+
+  public CokeOvenBlock(Properties properties) {
+    super(properties
+        .harvestTool(ToolType.PICKAXE));
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/CokeOvenBlock.java]: Move property definitions to where you register the block.

File: src/main/java/mods/railcraft/world/level/block/entity/CokeOvenBlockEntity.java [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+      return true;
+    }
+    return false;
+  }
+
+  @Override
+  public void load(BlockState blockState, CompoundNBT data) {
+    super.load(blockState, data);
+    this.items = NonNullList.withSize(this.getContainerSize(), ItemStack.EMPTY);
+    ItemStackHelper.loadAllItems(data, this.items);
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/entity/CokeOvenBlockEntity.java]: Use camel case for NBT tag names.

File: src/main/java/mods/railcraft/world/level/block/entity/ManualRollingMachineBlockEntity.java [hunk 4]
```diff
@@ -148,7 +148,14 @@ public void setItem(int slotID, ItemStack stack) {
 
   @Override
   public boolean stillValid(PlayerEntity playerEntity) {
-    return true;
+    if (this.level.getBlockEntity(this.worldPosition) != this) {
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/entity/ManualRollingMachineBlockEntity.java:151]: Just check `this.isRemoved()`

File: src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    //   this.verifyLink();
+    // }
+  }
+
+  @Override
+  public CompoundNBT save(CompoundNBT data) {
+    super.save(data);
+    data.putBoolean("isFormed", this.formed);
+    data.put("parentPos", NBTUtil.writeBlockPos(this.parentPos));
+    data.put("ourNormal", NBTUtil.writeBlockPos(this.normal));
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java]: Rename NBT key to normal

File: src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+    // if (this.isFormed()) {
+    //   this.verifyLink();
+    // }
+  }
+
+  @Override
+  public CompoundNBT save(CompoundNBT data) {
+    super.save(data);
+    data.putBoolean("isFormed", this.formed);
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java]: Rename NBT key to formed

File: src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+    }
+
+    if (this.parentPos.equals(BlockPos.ZERO)) {
+      return (T) this;
+    }
+    // it's totaly safe :) - speaking of!
+    // TODO: implement chunk safety
+    @Nullable TileEntity doesItExist;
+    try {
+      doesItExist = this.getLevel().getBlockEntity(parentPos);
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java]: This shouldn't ever throw an exception.

File: src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+    this.delink(); // iirc we should not have linked stuffed before
+    this.parentPos = parentPos;
+    this.formed = true;
+    this.setChanged();
+  }
+
+  /**
+   * .
+   */
+  public @Nullable T getParent() {
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java]: Move the annotation above the method.

File: src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+import net.minecraft.tileentity.TileEntityType;
+import net.minecraft.util.Direction;
+import net.minecraft.util.math.BlockPos;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class MultiblockEntity<T extends MultiblockEntity<T>> extends TileEntity {
+
+  public static final Logger MULTIBLOCK_LOGGER = LogManager.getLogger("Railcraft/MultiblockEntity");
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java]: Rename this to `logger`, use `LogManager.getLogger()` and make it private. (This applies to the other logger too)

File: src/main/java/mods/railcraft/world/level/block/entity/multiblock/CokeOvenMultiblockEntity.java [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+      this.setChanged();
+    }
+  }
+
+  @Override
+  public boolean stillValid(PlayerEntity playerEntity) {
+    if (!this.isFormed()) {
+      return false;
+    }
+    if (this.level.getBlockEntity(this.worldPosition) != this) {
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/entity/multiblock/CokeOvenMultiblockEntity.java]: Check `this.isRemoved()`

**Review by josephtarbit — CHANGES_REQUESTED** [review]

File: src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java (hunk 9 above)
Comment (LetterN) [reply src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java]: currently used for testing

File: src/main/java/mods/railcraft/api/signal/DualSignalReceiver.java [hunk 11]
```diff
@@ -9,7 +9,7 @@
 
 public class DualSignalReceiver extends SingleSignalReceiver {
 
-  private static final Logger logger = LogManager.getLogger();
+  private static final Logger logger = LogManager.getLogger("Railcraft/DualSignalReceiver");
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/api/signal/DualSignalReceiver.java]: Undo this.

File: src/main/java/mods/railcraft/client/gui/screen/inventory/CokeOvenMenuScreen.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+import mods.railcraft.Railcraft;
+import mods.railcraft.world.item.crafting.CokeOvenMenu;
+import net.minecraft.client.gui.screen.inventory.ContainerScreen;
+import net.minecraft.entity.player.PlayerInventory;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.text.ITextComponent;
+
+public class CokeOvenMenuScreen extends ContainerScreen<CokeOvenMenu> {
+  private static final ResourceLocation BACKGROUND_TEXTURE =
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/client/gui/screen/inventory/CokeOvenMenuScreen.java:14]: Add new line before this.

File: src/main/java/mods/railcraft/world/level/block/CokeOvenBlock.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+  protected void openContainer(World world, BlockPos blockPos,
+      PlayerEntity player, Direction facingDir) {
+    TileEntity blockEntity = world.getBlockEntity(blockPos);
+    if (!(blockEntity instanceof CokeOvenMultiblockEntity)) {
+      return;
+    }
+    CokeOvenMultiblockEntity recast = (CokeOvenMultiblockEntity) blockEntity;
+
+    if (!recast.isFormed()) {
+      if (!recast.tryToMakeParent(facingDir)) {
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/CokeOvenBlock.java]: Move this to the same `if`

File: src/main/java/mods/railcraft/world/level/block/entity/multiblock/CokeOvenMultiblockEntity.java [hunk 14]
```diff
@@ -0,0 +1,434 @@
...
+import net.minecraftforge.fluids.capability.IFluidHandler.FluidAction;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class CokeOvenMultiblockEntity extends MultiblockEntity<CokeOvenMultiblockEntity>
+    implements INamedContainerProvider, IRecipeHolder, IFluidTank, IInventory {
+
+  private static final Logger LOGGER =
+      LogManager.getLogger("Railcraft/MultiblockEntity/CokeOvenMultiblockEntity");
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/entity/multiblock/CokeOvenMultiblockEntity.java]: Use `LogManager.getLogger()`

File: src/main/java/mods/railcraft/world/level/block/entity/multiblock/CokeOvenMultiblockEntity.java [hunk 15]
```diff
@@ -0,0 +1,434 @@
...
+import net.minecraftforge.fluids.IFluidTank;
+import net.minecraftforge.fluids.capability.IFluidHandler.FluidAction;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class CokeOvenMultiblockEntity extends MultiblockEntity<CokeOvenMultiblockEntity>
+    implements INamedContainerProvider, IRecipeHolder, IFluidTank, IInventory {
+
+  private static final Logger LOGGER =
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/entity/multiblock/CokeOvenMultiblockEntity.java]: Use lower case.

File: src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+import net.minecraft.util.math.BlockPos;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class MultiblockEntity<T extends MultiblockEntity<T>> extends TileEntity
+    implements ITickableTileEntity {
+
+  private static final Logger LOGGER =
+      LogManager.getLogger("Railcraft/MultiblockEntity");
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java:24]: Do the same as what I said for the other logger 😉

File: src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+    // do not save the normal when it's zero.
+    if (this.normal == BlockPos.ZERO) {
+      data.put("normal", NBTUtil.writeBlockPos(this.normal));
+    }
+    return data;
+  }
+
+  @Override
+  public void tick() {
+    if (!this.initialized && this.shouldInitialize) {
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/entity/multiblock/MultiblockEntity.java]: Extend `RailcraftTickableBlockEntity` and override `load` (it does what you've written for you)

File: src/main/resources/assets/railcraft/lang/en_us.json [hunk 18]
```diff
@@ -234,6 +235,7 @@
     "season.none": "None",
 
     "container.manual_rolling_machine": "Manual Rolling Machine",
+    "container.coke_oven_multiblock": "Coke Oven",
```
Comment (josephtarbit) [inline src/main/resources/assets/railcraft/lang/en_us.json]: Rename the lang key to just container.coke_oven

File: src/main/java/mods/railcraft/world/level/block/RailcraftBlocks.java [hunk 19]
```diff
@@ -586,6 +586,12 @@
           () -> new ManualRollingMachineBlock(AbstractBlock.Properties.of(Material.WOOD)
               .sound(SoundType.WOOD)));
 
+  public static final RegistryObject<Block> COKE_OVEN_BLOCK =
+      BLOCKS.register("coke_oven_block",
```
Comment (josephtarbit) [inline src/main/java/mods/railcraft/world/level/block/RailcraftBlocks.java]: Rename to just coke_oven

**Review by josephtarbit — CHANGES_REQUESTED** [review]

**Pull request merged**